### PR TITLE
use golang proto for alpn filter

### PR DESCRIPTION
**Pull request by yxue**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by yxue** [comment]
cc @kyessenov

File: security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2018 Istio Authors
```
Comment (kyessenov) [inline security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto]: 2019

File: security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+syntax = "proto3";
+
+// $title: ALPN filter for overriding ALPN for upstream TLS connections.
+
+package istio.envoy.config.filter.http.alpn.v2alpha1;
```
Comment (kyessenov) [inline security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto]: why v2alpha1? just v1?

Comment (yxue) [reply security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto]: Actually I don't know which version I should choose. Just follow the pattern from existing filter, e.g. tcp_cluster_rewrite filter

**Comment by yxue** [comment]
> How is this filter implemented? Is this an Istio filter or an Envoy upstream filter?

It's a HTTP filter which will be in istio/proxy since currently only istio uses it.

> Asking because the upstream socket configuration can be either `cluster.transport_socket` or `cluster.tls_context` or `cluster.transport_socket_match`

The filter will overrides the ALPN in ssl transport socket. https://github.com/envoyproxy/envoy/blob/3582d02f6600057e676629c2afefe5265e669b10/source/extensions/transport_sockets/tls/context_impl.cc#L809 I don't think it related with how transport socket is configured.

**Comment by incfly** [comment]
Okay sounds good, so it's one layer below the Cluster API, should be fine.

File: security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto (hunk 2 above)
Comment (incfly) [reply security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto]: that seems a mistake of that filter/weirdness though, without much context...

File: security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+option go_package = "istio.io/api/envoy/config/filter/http/alpn/v2alpha1";
+
+// FilterConfig is the config for Istio-specific filter.
```
Comment (incfly) [inline security/proto/envoy/config/filter/http/alpn/v2alpha1/config.proto]: s/`Istio-specific filter` .e.g Istio ALPN rewrite filter/g ?

**Comment by yxue** [comment]
@istio/wg-security-maintainers could you help to take a look?

**Comment by kyessenov** [comment]
BTW we don't want any custom filters long-term. So please make sure to take a task to upstream whatever customizations are needed or use WASM ABI to implement it.

**Comment by yxue** [comment]
> BTW we don't want any custom filters long-term. So please make sure to take a task to upstream whatever customizations are needed or use WASM ABI to implement it.

FYI https://github.com/envoyproxy/envoy/pull/8498

I created the filter in Envoy upstream at the very beginning. Envoy community would like to accept the change until other users require it. We might be able to implement it using WASM.

**Pull request merged**